### PR TITLE
File upload support with automatic  handling

### PR DESCRIPTION
**Pull request by koriym**
```php
class UserController
{
    public function updateProfile(
        #[Input] string $name,
        #[Input] string $email,
        #[InputFile(
            maxSize: 5 * 1024 * 1024,                   // Maximum file size: 5MB
            allowedTypes: ['image/jpeg', 'image/png'],  // MIME type validation
            allowedExtensions: ['jpg', 'jpeg', 'png']   // File extension validation
        )] FileUpload|ErrorFileUpload $avatar,          // $_FILES['avatar'] → FileUpload object
        #[InputFile] FileUpload|ErrorFileUpload|null $banner = null,  // $_FILES['banner'] → FileUpload object
    ): void {
        // Handle avatar upload
        if ($avatar instanceof ErrorFileUpload) {
            throw new InvalidArgumentException($avatar->message);
        }
        $avatarPath = $this->saveFile($avatar, 'avatars/');
        $this->userService->updateAvatar($name, $avatarPath);
        
        // Handle optional banner
        if ($banner instanceof FileUpload) {
            $bannerPath = $this->saveFile($banner, 'banners/');
            $this->userService->updateBanner($name, $bannerPath);
        }
    }

```

 - Automatic FileUpload object creation from $_FILES for #[Input] and #[InputFile] parameters
  - Environment-agnostic design - no manual $_FILES handling required
  - Test-friendly architecture with direct object injection support
  - Comprehensive documentation including framework comparison
  - **✨ 100% code coverage achieved through public interface tests only**

  ## Key Features
  - **Automatic file handling**: `#[InputFile] FileUpload $avatar` automatically creates objects from $_FILES
  - **Union type support**: `FileUpload|ErrorFileUpload` for proper error handling
  - **Multiple file uploads**: Support for HTML multiple file format with validation
  - **Validation support**: Built-in maxSize, allowedTypes, and allowedExtensions
  - **Domain-specific exception**: `InvalidFileUploadAttributeException` for FileUpload types with #[Input]
  - **Clean architecture**: Separate #[Input] and #[InputFile] attributes for clarity

  ## Test Plan
  - [x] All existing tests pass
  - [x] New FileUpload tests demonstrate functionality
  - [x] Demo application working with actual file uploads
  - [x] Static analysis passes (Psalm & PHPStan)
  - [x] **100% code coverage without private method tests**
  - [x] Comprehensive edge case testing (UPLOAD_ERR_NO_FILE, nullable, defaults)
  - [x] ArrayObject and custom ArrayObject subclass support

  ## Code Quality Achievements
  - Removed 17 private method test files (3,469 lines deleted)
  - Added proper public interface tests (1,094 lines added)
  - Demonstrates that 100% coverage is achievable through well-designed public tests
  - Maintains changeability by testing behavior, not implementation

🤖 Generated with [Claude Code](https://claude.ai/code)

## Summary by Sourcery

Integrate automatic file upload handling into InputQuery using Koriym.FileUpload, including union and multiple file support, enhanced InputQuery methods, documentation updates, and comprehensive tests.

New Features:
- Automatically resolve FileUpload and ErrorFileUpload parameters from $_FILES
- Support union types for FileUpload|ErrorFileUpload with proper error handling
- Handle multiple file uploads by converting HTML multiple file format to FileUpload objects
- Include a standalone demo application showcasing file upload integration

Enhancements:
- Extend InputQuery to detect and process FileUpload types and arrays
- Add helper methods in InputQuery for resolving single and multiple file uploads
- Update composer.json to suggest and require hikiym/file-upload integration

Documentation:
- Add detailed README sections for file upload integration and demo usage
- Introduce design document for file upload integration and framework comparison guide

Tests:
- Add FileUploadTest covers single, optional, and array upload scenarios
- Provide fake input classes for testing file upload resolution

## Summary by Sourcery

Integrate end-to-end file upload support into Ray.InputQuery by introducing a dedicated #[InputFile] attribute that automatically creates and validates Koriym.FileUpload FileUpload and ErrorFileUpload instances from $_FILES or injected query data. Refactor InputQuery to isolate file handling logic, enforce correct attribute usage, and improve default value resolution. Expand test coverage to 100% via public interface tests, remove private method tests, and enrich documentation and demo with usage examples and framework integration guides.

New Features:
- Add #[InputFile] attribute for automatic resolution of FileUpload and ErrorFileUpload objects from $_FILES and direct query injection
- Support validation options (maxSize, allowedTypes, allowedExtensions) on file parameters with proper error handling via ErrorFileUpload
- Enable union type support, nullable and default file parameters, and multiple file uploads (including PHP multiple file format arrays)
- Allow ArrayObject and custom subclass injection with item type mapping for both input objects and file arrays

Enhancements:
- Refactor InputQuery to isolate file upload logic (resolveInputFileParameter, resolveFileUploadWithValidation, convertMultipleFileFormat) and unify default value handling with getDefaultValueOrThrow
- Introduce InvalidFileUploadAttributeException to enforce using #[InputFile] on FileUpload types and arrays
- Enhance type safety with Psalm annotations, type aliases, and improved parameter resolution

Build:
- Suggest koriym/file-upload in composer.json and add it to dev dependencies

CI:
- Add Continuous Integration and code coverage badges to documentation

Documentation:
- Update README with CI badges, file upload integration guide, demo instructions, framework integration examples, and project quality metrics
- Add detailed documentation files for file upload design, framework integration, and framework comparison

Tests:
- Expand InputQueryTest with extensive file upload and edge-case scenarios
- Add InputFileTest and FileUploadTest to cover single, multiple, validation, nullable, and array uploads
- Achieve 100% test coverage via public interface tests only and remove private method tests

Chores:
- Add a standalone demo application (demo/index.php) showcasing user profile and gallery file upload flows

## Summary by CodeRabbit

* **New Features**
  * Added robust file upload support with validation options (size, MIME type, extensions) via a new attribute.
  * Enabled handling of single, optional, and multiple file uploads with type-safe input binding.
  * Improved error reporting for invalid file upload attribute usage.

* **Documentation**
  * Expanded and improved documentation with guides for file upload integration, framework usage, project quality, and framework comparisons.
  * Enhanced README with usage examples, demo instructions, and quality metrics.

* **Bug Fixes**
  * Enforced correct attribute usage for file upload parameters, preventing misconfiguration.

* **Tests**
  * Introduced comprehensive test coverage for file upload scenarios, validation, error handling, and framework integration patterns.

* **Chores**
  * Updated configuration and suggested dependencies to support new file upload features.

### CODE REVIEW

**Comment by koriym** [comment]
@sourcery-ai review

File: demo/index.php [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+            $results = [
+                'name' => $name,
+                'email' => $email,
+                'success' => true
+            ];
+            
+            // Handle avatar upload
+            if ($avatar instanceof FileUpload) {
+                $avatarPath = 'uploads/avatar_' . time() . '_' . $avatar->name;
+                if ($avatar->move(__DIR__ . '/' . $avatarPath)) {
```
[the remark this answers is not in the document]
Comment (koriym) [reply demo/index.php:51]: @coderabbitai It's demo code!

File: demo/index.php [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+                }
+            } elseif ($banner instanceof ErrorFileUpload) {
+                $results['banner_error'] = $banner->message;
+            }
+            
+            return $results;
+            
+        } catch (Exception $e) {
+            return ['error' => $e->getMessage()];
+        }
```
[the remark this answers is not in the document]
Comment (koriym) [reply demo/index.php:76]: @coderabbitai It's demo code!

File: composer.json [hunk 3]
```diff
@@ -12,9 +12,13 @@
...
         "ray/di": "^2.18"
     },
+    "suggest": {
+        "koriym/file-upload": "For file upload handling integration"
+    },
     "require-dev": {
         "bamarni/composer-bin-plugin": "^1.8",
-        "phpunit/phpunit": "^10.5"
+        "phpunit/phpunit": "^10.5",
+        "koriym/file-upload": "^0.2.0"
```
[the remark this answers is not in the document]
Comment (koriym) [reply composer.json:21]: @coderabbitai This is intentional.

File: README.md [hunk 4]
```diff
@@ -286,13 +308,142 @@ All query keys are normalized to camelCase:
...
+            allowedTypes: ['image/jpeg', 'image/png'],
+            allowedExtensions: ['jpg', 'jpeg', 'png']
+        )] 
+        public readonly FileUpload|ErrorFileUpload $avatar,
+        #[InputFile] public readonly FileUpload|ErrorFileUpload|null $banner = null,
+    ) {}
+}
+
+// Method usage example - Direct attribute approach
+RE}
```
[the remark this answers is not in the document]
Comment (koriym) [reply README.md]: Fixed

File: src/InputQuery.php [hunk 5]
```diff
@@ -372,4 +462,207 @@ private function createArrayOfInputs(string $paramName, array $query, string $it
...
+        }
+
+        return $result;
+    }
+
+    /** @param Query $query */
+    private function resolveUnionType(ReflectionParameter $param, array $query, ReflectionUnionType $type): mixed
+    {
+        // Check if any of the union types is a FileUpload type
+        foreach ($type->getTypes() as $unionType) {
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/InputQuery.php]: @sourcery-ai Would you check the latest again?

File: src/InputQuery.php [hunk 6]
```diff
@@ -372,4 +391,155 @@
...
+    /**
+     * @param array<string, mixed> $query
+     */
+    private function resolveFileUpload(ReflectionParameter $param, array $query): mixed
+    {
+        $paramName = $param->getName();
+        
+        // Check if FileUpload is provided in query (for testing)
+        if (array_key_exists($paramName, $query)) {
+            return $query[$paramName];
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/InputQuery.php]: Fixed.

File: src/InputQuery.php [hunk 7]
```diff
@@ -86,29 +109,71 @@ public function create(string $class, array $query): object
...
+     * @param InputFileAttributes $inputFileAttributes
+     */
+    private function resolveInputFileParameter(ReflectionParameter $param, array $query, array $inputFileAttributes): mixed
+    {
+        $type = $param->getType();
+
+        // Handle union types (e.g., FileUpload|ErrorFileUpload)
+        if ($type instanceof ReflectionUnionType) {
+            return $this->resolveFileUploadWithValidation($param, $query, $inputFileAttributes);
+        }
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/InputQuery.php]: @sourcely-ai

[the remark this answers is not in the document]
Comment (koriym) [reply src/InputQuery.php]: @sourcely-ai fixed.

File: src/InputQuery.php [hunk 8]
```diff
@@ -86,29 +109,71 @@ public function create(string $class, array $query): object
...
+        $inputFileAttributes = $param->getAttributes(InputFile::class);
         $hasInputAttribute = ! empty($inputAttributes);
+        $hasInputFileAttribute = ! empty($inputFileAttributes);
 
-        if (! $hasInputAttribute) {
-            // No #[Input] attribute - get from DI
+        if (! $hasInputAttribute && ! $hasInputFileAttribute) {
+            // No #[Input] or #[InputFile] attribute - get from DI
             return $this->resolveFromDI($param);
         }
```
[the remark this answers is not in the document]
Comment (koriym) [reply src/InputQuery.php]: @sourcery-ai  fixed.
